pkg/util/validation/limits.go

```diff
@@ -73,7 +76,177 @@ func (l *Limits) UnmarshalYAML(unmarshal func(interface{}) error) error {
+func (o *Overrides) CreationGracePeriod(userID string) time.Duration {
+	return o.overridesManager.GetLimits(userID).(*Limits).CreationGracePeriod
+}
+
+// MaxSeriesPerQuery returns the maximum number of series a query is allowed to hit.
+func (o *Overrides) MaxSeriesPerQuery(userID string) int {
+	return o.overridesManager.GetLimits(userID).(*Limits).MaxSeriesPerQuery
+}
+
+// MaxSamplesPerQuery returns the maximum number of samples in a query (from the ingester).
+func (o *Overrides) MaxSamplesPerQuery(userID string) int {
+	return o.overridesManager.GetLimits(userID).(*Limits).MaxSamplesPerQuery
+}
+
+// MaxSeriesPerUser returns the maximum number of series a user is allowed to store.
+func (o *Overrides) MaxSeriesPerUser(userID string) int {
+	return o.overridesManager.GetLimits(userID).(*Limits).MaxSeriesPerUser
+}
+
+// MaxSeriesPerMetric returns the maximum number of series allowed per metric.
+func (o *Overrides) MaxSeriesPerMetric(userID string) int {
+	return o.overridesManager.GetLimits(userID).(*Limits).MaxSeriesPerMetric
+}
+
+// MaxChunksPerQuery returns the maximum number of chunks allowed per query.
+func (o *Overrides) MaxChunksPerQuery(userID string) int {
+	return o.overridesManager.GetLimits(userID).(*Limits).MaxChunksPerQuery
+}
+
+// MaxQueryLength returns the limit of the length (in time) of a query.
+func (o *Overrides) MaxQueryLength(userID string) time.Duration {
+	return o.overridesManager.GetLimits(userID).(*Limits).MaxQueryLength
+}
+
+// MaxQueryParallelism returns the limit to the number of sub-queries the
+// frontend will process in parallel.
+func (o *Overrides) MaxQueryParallelism(userID string) int {
+	return o.overridesManager.GetLimits(userID).(*Limits).MaxQueryParallelism
+}
+
+// EnforceMetricName whether to enforce the presence of a metric name.
+func (o *Overrides) EnforceMetricName(userID string) bool {
+	return o.overridesManager.GetLimits(userID).(*Limits).EnforceMetricName
+}
+
+// CardinalityLimit whether to enforce the presence of a metric name.
+func (o *Overrides) CardinalityLimit(userID string) int {
+	return o.overridesManager.GetLimits(userID).(*Limits).CardinalityLimit
+}
+
+// Loads overrides and returns the limits as an interface to store them in OverridesManager.
+// We need to implement it here since OverridesManager must store type Limits in an interface but
+// it doesn't know its definition to initialize it.
+func loadOverrides(filename string) (map[string]interface{}, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	var overrides struct {
+		Overrides map[string]*Limits `yaml:"overrides"`
+	}
+
+	decoder := yaml.NewDecoder(f)
+	decoder.SetStrict(true)
+	if err := decoder.Decode(&overrides); err != nil {
+		return nil, err
+	}
+
+	overridesAsInterface := map[string]interface{}{}
+	for userID := range overrides.Overrides {
+		overridesAsInterface[userID] = overrides.Overrides[userID]
+	}
+
+	return overridesAsInterface, nil
+}
```

Comment:
Can we make this take a "factory" for the `Limits` type?  Then have the code to make a map vs the defaults would be in override.go.

I think this is possible with `yaml.v3` and `Node`: https://godoc.org/gopkg.in/yaml.v3

Reply:
As discussed offline, `Node.Decode` doesn't honour `KnownFields`: https://github.com/go-yaml/yaml/issues/460